packages/typescript-estree/src/parser-options.ts

```diff
@@ -182,12 +182,21 @@ export interface ParserWeakMapESTreeToTSNode<
   has(key: unknown): boolean;
 }
 
-export interface ParserServices {
-  program: ts.Program;
+export interface ParserServicesNodeMaps {
   esTreeNodeToTSNodeMap: ParserWeakMapESTreeToTSNode;
   tsNodeToESTreeNodeMap: ParserWeakMap<TSNode | TSToken, TSESTree.Node>;
-  hasFullTypeInformation: boolean;
 }
+export interface ParserServicesWithTypeInformation
+  extends ParserServicesNodeMaps {
+  program: ts.Program;
+}
+export interface ParserServicesWithoutTypeInformation
+  extends ParserServicesNodeMaps {
+  program: null;
```

Comment:
Seems odd to me that we'd have to explicitly add `program: null`. How about `?: never`?

```suggestion
  program?: never;
```

Here and in `shared.ts` > `interface ASTAndNoProgram`.

Reply:
IMO it's nice to have an explicit and known value for it. "it's always there and it's either the program or it's null" is explicit and easy to reason about.

It's the same thing I'm trying to do with the AST - the properties always exist and they have explicit empty states.